dali/core/os/shared_mem.cc

```diff
@@ -0,0 +1,164 @@
+  POSIX_CALL_EX(unlink(temp_path), "Couldn't unlink temporary file.");
+  const char *temp_filename = temp_path + shm_size;
+  shm_handle_t fd = shm_open(temp_filename, O_CREAT | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+  POSIX_CHECK_STATUS(temp_fd, "shm_open");
+  ::close(temp_fd);
+  // The memory chunks will be passed between processes by sending descriptors over the sockets,
+  // not through the filenames; this way in case of unexpected process exit we're not leaking any
+  // filelike objects in /dev/shm or /tmp. The OS guarantees to keep the memory as long as
+  // any process has fds or has mmaped the memory.
+  shm_unlink(temp_filename);
+  return ShmHandle(fd);
+}
+
+void ShmHandle::DestroyHandle(shm_handle_t h) {
+  if (h >= 0) {
+    POSIX_CALL(::close(h));
+  }
+}
+
+MemoryMapping::MemoryMapping(shm_handle_t handle, uint64_t size)
+    : UniqueHandle<MappedMemoryChunk, MemoryMapping>(
+          {size, static_cast<uint8_t *>(
+                     mmap(nullptr, size, PROT_WRITE | PROT_READ, MAP_SHARED, handle, 0))}) {
+  if (handle_.ptr == MAP_FAILED) {
+    POSIX_CHECK_STATUS(-1, "mmap");
+  }
+}
+
+void MemoryMapping::DestroyHandle(MappedMemoryChunk handle) {
+  if (handle.ptr) {
+    POSIX_CALL(munmap(handle.ptr, handle.size));
+  }
+}
+
+uint8_t *MemoryMapping::get_raw_ptr() {
+  return handle_.ptr;
+}
+
+
+void MemoryMapping::resize(uint64_t new_size) {
+  if (!handle_.ptr) {
+    throw std::runtime_error("Cannot resize the memory mapping, because no memory is mapped.");
+  }
+  handle_.ptr = static_cast<uint8_t *>(mremap(handle_.ptr, handle_.size, new_size, MREMAP_MAYMOVE));
+  if (handle_.ptr == MAP_FAILED) {
+    POSIX_CHECK_STATUS(-1, "mremap");
+  }
+  handle_.size = new_size;
+}
+
+SharedMem::SharedMem(shm_handle_t handle, uint64_t size) : size_{size * sizeof(SharedMem::b_type)} {
+  if (handle >= 0) {
+    shm_handle_ = ShmHandle(handle);
+  } else {
+    shm_handle_ = ShmHandle::CreateHandle();
+    POSIX_CALL_EX(ftruncate(shm_handle_, size_), "Failed to resize shared memory.");
+  }
+  memory_mapping_ = MemoryMapping(shm_handle_, size_);
+}
+
+uint64_t SharedMem::size() const {
+  return size_;
+}
+
+int SharedMem::handle() {
+  return !shm_handle_ ? -1 : shm_handle_;
+}
+
+SharedMem::b_type *SharedMem::get_raw_ptr() {
+  return !memory_mapping_ ? nullptr : memory_mapping_.get_raw_ptr();
+}
+
+void SharedMem::resize(uint64_t size, bool trunc) {
+  size_ = size * sizeof(SharedMem::b_type);
+  if (trunc) {
+    POSIX_CALL_EX(ftruncate(shm_handle_, size_), "Failed to resize shared memory.");
+  }
+  if (memory_mapping_) {
+    memory_mapping_.resize(size_);
+  } else {
+    if (!shm_handle_) {
+      throw std::runtime_error("Cannot mmap memory - no valid shared memory handle.");
+    }
+    memory_mapping_ = MemoryMapping(shm_handle_, size_);
+  }
+}
+
+void SharedMem::close() {
+  if (memory_mapping_) {
+    memory_mapping_.reset();
+  }
+  if (shm_handle_) {
+    shm_handle_.reset();
+  }
```

Comment:
This `if` is already there in `reset`.
```suggestion
  memory_mapping_.reset();
  shm_handle_.reset();
```

Reply:
done